Games/types/Mafia/items/Gun.js

```diff
@@ -10,53 +10,75 @@ module.exports = class Gun extends Item {
-
-                        if (cursed) {
-                            this.target = this.actor;
-                        }
-
-                        if (reveal && cursed)
-                            this.game.queueAlert(`:sy0b: ${shooterMask} pulls a gun, it backfires!`);
-                        else if (reveal && !cursed)
-                            this.game.queueAlert(`:sy0a: ${shooterMask} pulls a gun and shoots at ${this.target.name}!`);
-                        else
-                            this.game.queueAlert(`:sy0a: Someone fires a gun at ${this.target.name}!`);
-
-                        // kill
-                        var toKill = this.dominates();
-                        if (mafiaImmune && this.target.role.alignment == "Mafia")
-                            toKill = false
-
-                        if (toKill) {
-                            this.target.kill("gun", this.actor, true);
-                        }
-
-                        this.item.drop();
-                        this.game.broadcast("gunshot");
+        this.baseMeetingName = "Shoot Gun";
+        this.currentMeetingIndex = 0;
+
+        this.gunMeeting = {
+            actionName: "Shoot",
+            states: ["Day"],
+            flags: ["voting", "instant", "noVeg"],
+            action: {
+                labels: ["kill", "gun"],
+                item: this,
+                run: function () {
+                    var shooterMask = this.actor.role.data.shooterMask;
+                    var reveal = shooterMask ? true : this.item.reveal;
+                    if (reveal == null) {
+                        reveal = Random.randArrayVal([true, false]);
+                    }
+                    if (shooterMask == null) {
+                        shooterMask = this.actor.name;
+                    }
+                    
+                    var mafiaImmune = this.item.mafiaImmune;
+                    var cursed = this.item.cursed;
+
+                    if (cursed) {
+                        this.target = this.actor;
+                    }
+
+                    if (reveal && cursed)
+                        this.game.queueAlert(`:sy0b: ${shooterMask} pulls a gun, it backfires!`);
+                    else if (reveal && !cursed)
+                        this.game.queueAlert(`:sy0a: ${shooterMask} pulls a gun and shoots at ${this.target.name}!`);
+                    else
+                        this.game.queueAlert(`:sy0a: Someone fires a gun at ${this.target.name}!`);
+
+                    // kill
+                    var toKill = this.dominates();
+                    if (mafiaImmune && this.target.role.alignment == "Mafia")
+                        toKill = false
+                    
+                    let toSnatch = false;
+                    // snatch gun
+                    if (!cursed && 
+                        (this.target.role.name == "Gunslinger" || this.target.role.name == "Sharpshooter") &&

```

Comment:
Gun immunity should be a feature of the roles not the gun

Reply:
the main reason of this design is to snatch back the exact same gun instead of creating a new gun item

Reply:
I'll go back and try using the instant action listener to implement this

Reply:
just use the "immune" listener

Reply:
ok I'll check that, should it regenerate a new gun or use back the old gun

Reply:
It should be a new gun, I'd say if it's cursed then just kill

Reply:
i made the changes, going to test later tonight
i want to keep the same gun is because snatching asso gun should also not kill mafia after snatching

Reply:
illu gun shouldnt do the illu thing if someone else uses it

Reply:
ok i'll test that later, i believe it won't do that rn because the shooterMask checks actor.role.data